Installing debian packages in the armbian build system

The-going opened the pull request:
Description

 In many places, we use the same mechanism for installing a series
 of packages, checking each time which of these packages are already
 installed. And we do this action a little differently each time.
 And we do not always ensure the correct output of errors in the logs.
    
    This function is an attempt to optimize and unify the code.
    Let's use it to install a series of packages.

@igorpecovnik @lanefu Before I post the next part, I want to hear the opinion of other team members about this.

igorpecovnik commented:
I think this is a good idea to clean up and merge different styles of approaching a problem. Does it make sense to expand it to chroot functionality https://github.com/armbian/build/blob/master/lib/image-helpers.sh#L180-L205

The-going commented:
> I think this is a good idea to clean up and merge different styles of approaching a problem. Does it make sense to expand it to chroot functionality https://github.com/armbian/build/blob/master/lib/image-helpers.sh#L180-L205

I currently have two options for the install_deb_chroot function. The first is a wrapper that creates a temporary executable script with the install_pkg_deb function with a list of packages and executing it. And running this script in a chroot environment. The second is an extension of the existing install_deb_chroot script to the install_pkg_deb functionality. In any case, two functions are obtained. One for native installation and the other for installation in a chroot environment.

igorpecovnik commented:
> In any case, two functions are obtained. One for native installation and the other for installation in a chroot environment.

Yeah, better. OK. I am fine with merging this one.

The-going commented:
In a few days I will make some commits with a working use case.

igorpecovnik commented:
If you are going to add to this commit, pick a label WIP.

The-going commented:
> If you are going to add to this commit, pick a label WIP.

Yes, I agree this is WIP. I found a problem. The APT cache and the DPKG cache are two different caches. The package may not be in the dpkg cache at all, but the APT knows about the package. The function needs to be improved.
I don't have much free time right now. It is necessary to harvest berries: raspberries, currants, gooseberries ... So everything else is slow.

The-going commented:
The install_pkg_deb function works well and has been tested.
There are still questions about the script chroot-buildpackages.sh, but this has nothing to do with the function itself and it can be discussed. Please write your opinion.
For example, for a hostapd-realtek package, there is a serious risk of confusion for calculating dependencies for building a package.
Building only this one package in a clean environment will cause a crash.
We take the package_builddeps variable from [this location](https://github.com/armbian/build/blob/8d64acfd21d6885c1179d824777aa6125cdc078e/packages/extras-buildpkgs/90-hostapd-realtek.conf#L6).
But all the dependencies for the build are spelled out correctly in the [hostapd-realtek/debian/control](https://github.com/armbian/build/blob/8d64acfd21d6885c1179d824777aa6125cdc078e/packages/extras-buildpkgs/hostapd-realtek/debian/control#L8) file.
Here is my speech about [this library](https://github.com/armbian/build/blob/8d64acfd21d6885c1179d824777aa6125cdc078e/packages/extras-buildpkgs/hostapd-realtek/debian/control#L10).

If I changed [the code](https://github.com/armbian/build/blob/8d64acfd21d6885c1179d824777aa6125cdc078e/lib/chroot-buildpackages.sh#L296) this way:

```
	cd /root/build/"${package_name}"
	# copy overlay / "debianization" files
	[[ -d "/root/overlay/${package_name}/" ]] && rsync -aq /root/overlay/"${package_name}" /root/build/

        package_builddeps="\$(dpkg-checkbuilddeps |& awk -F":" '{print \$NF}')"
        if [[ -n "\${package_builddeps}" ]]; then
                install_pkg_deb \${package_builddeps}
        fi

```

then the build will be successful.

The-going commented:
This PR has logically ended.
The result of her work allowed us to make the last commit.

igorpecovnik commented:
Ready for review and merge if everything is ok?

lanefu commented:
@igorpecovnik  can you refresh RC before merging this?  I think this change is probably the cutoff for Release vs master

igorpecovnik commented:
OK, we can start with cherrypicking from now on then.

The-going commented:
> Does it make sense to expand it to chroot functionality https://github.com/armbian/build/blob/master/lib/image-helpers.sh#L180-L205

I think that this function can be completely abandoned if we make the [installation only from the repositories](https://forum.armbian.com/topic/18330-armbian-2108-caracal-release-thread/?do=findComment&comment=126228)

The-going commented:
In my configuration, only these packages were collected:
```
~/build$ ls output/debs/extra/focal-utils/
hostapd_2.9-102~armbian21.08.0-trunk+1_arm64.deb
hostapd-realtek_2.5-4~armbian21.08.0-trunk+1_arm64.deb
htop_2.2.0-5~armbian21.08.0-trunk+1_arm64.deb
mmc-utils_0~gita3d3331-3~armbian21.08.0-trunk+1_arm64.deb
sunxi-tools_1.4.2-2~armbian21.08.0-trunk+1_arm64.deb
```

The-going commented:
> lanefu marked this pull request as draft 3 months ago 
@lanefu Please tell me, what is it about this job that has alerted you? What am I doing wrong or have I missed potential risks in the future?

igorpecovnik commented:
I think because we didn't want to push this in before the (previous) release. Let's move it to review state and move on.

The-going commented:
> I think because we didn't want to push this in before the (previous) release. Let's move it to review state and move on.

Ok! Here I mixed two different works. Added a new feature and changed the method of installing dependencies to build some packages. In order for the PR to become complete, I need to complete this for all packages.

The-going commented:
Let it remain a template. When I finish this work, then I will divide it into two or three logical PR.

The-going commented:
@igorpecovnik  Please take a look at this commit https://github.com/armbian/build/pull/3007/commits/a3523b6e96281c0f6686369e60337c330fe0886c

Does it make sense to make this default behavior?
That is, to make a clean version of the environment and store it in the archive.
And deploy it before starting work. Archive size =~ 125 Mb

lanefu commented on lib/chroot-buildpackages.sh:
```diff
@@ -86,9 +103,20 @@ create_chroot()
+
+	case $release in
+		bullseye|focal|hirsute|sid)
+			chroot "${target_dir}" /bin/bash -c "ln -s /usr/bin/python3 /usr/bin/python"
```
Using this package would be a more appropriate solution

https://packages.debian.org/sid/main/python-is-python3

The-going replied:
There is no doubt that this is the best solution.
Thanks.

lanefu commented on lib/chroot-buildpackages.sh:
```diff
@@ -149,8 +181,29 @@ chroot_build_packages()
 
+			if [ ! -f "${target_dir}.tar.xz" ]; then
+				display_alert "Create a clean Environment archive" "${t_name}.tar.xz" "info"
+				local tmp_dir=$(mktemp -d /tmp/debootstrap-XXXXX)
```
Mktemp should stay within the Armbian work directory

`${SRC}/.tmp/`

The-going replied:
Yes. I agree.

lanefu commented:
@The-going  what's left on this before ready to merge?

The-going commented:
I propose to divide this PR template into two parts.
The first part is a new function for installing debian packages and an example of its use when building in a chroot environment. The second part is to change the methodology\logic of the assembly itself in the chroot environment.
The idea is to keep the build environment always clean in the archive and deploy it immediately before building each package.
In this I need the explicit consent of the participants of the armbian or arguments to refuse it.

lanefu commented:
@rpardini @igorpecovnik 

Thoughts?

I just hate having open PRs for months.

The-going commented:
Hate makes it difficult to think logically.
Just tell me what to do with it, provided that the work is not finished.

igorpecovnik commented:
I did quick scan and that code cleaning part is harmless anyway, the rest I would like to review, perhaps test. But need some larger space of time then I have right now.

lanefu commented:
> Hate makes it difficult to think logically.
> Just tell me what to do with it, provided that the work is not finished.

Makes it easy to think strategically..

How about Fix the 2 suggestions I made, and let's merge the code cleaning part, and spit the rest for a separate PR?

The-going commented:
> code cleaning part

Do you mean the last commit? https://github.com/armbian/build/pull/3007/commits/011ad017b92f4fa9984ff5f8e5659fbb41a8750e

The-going commented:
> Fix the 2 suggestions I made

It's done, but I haven't pushed it out yet. I need to rebase the branch to remove the debugging code. And there is no need to test the branch in this form. Wait a day or two.

The-going commented:
Closed for old age.
